cirq/testing/order_tester.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A utility class for testing ordering methods.
+
+To test an ordering method, create an OrderTester and add several
+equivalence groups or items to it. The order tester will check that
+the items within each group are all equal to each other, and every new
+added item or group is strictly ascending with regard to the previously
+added items or groups.
+
+It will also check that a==b implies hash(a)==hash(b).
+"""
+
+from typing import Any
+
+from cirq.testing.equals_tester import EqualsTester
+
+
+class OrderTester:
+    """Tests ordering against user-provided disjoint ordered groups or items."""
+
+    def __init__(self):
+        self.eq_tester = EqualsTester()
+        self.eq_tester.groups = []  # otherwise ClassUnknownToSubjects
+                                    # throws TypeError for comparisons
+
+    def _groups(self):
+        return self.eq_tester.groups
+
+    def _try_comparison(self, comp):
+        result = NotImplemented
+        try:
+            result = comp()
+        except TypeError as inst:
+            assert False, inst
+        return result
+
+    def _do_assert(self, one, two, symbol):
+
+        condition = one or two  # bool(NotImplemented) == True
+        message = "{} is inconsistent: {!r},{!r}".format(symbol, one, two)
+
+        assert condition, message
+
+
+    def _verify_ascending(self, v1, v2):
+        """Verifies that (v1, v2) is a strictly ascending sequence."""
+
+        lt_1 =  self._try_comparison(lambda: v1 < v2)
```

Comment:
Define a list of the six comparison operators with names and lambdas at the start of the file, then iterate through them here. Once with `v1` as the first item, and once with `v2` as the first item. Assert that the result on v1 vs v2 should match the result on int 1 vs int 2.

```python
comparisons = [
    ('<', lambda a, b: a < b),
    ...
]
```

```python
for s, f in comparisons:
    assert f(v1, v2) == f(1, 2), '{} {} {} returned {} instead of {}'.format(v1, s, v2, f(v1, v2), f(1, 2))
```

Reply:
You mean four?
<, >, <=, >=